lightning/src/offers/async_receive_offer_cache.rs

```diff
@@ -0,0 +1,493 @@
+	///
+	// We need to re-persist the cache if a fresh offer was just marked as used to ensure we continue
+	// to keep this offer's invoice updated and don't replace it with the server.
+	pub fn get_async_receive_offer(
+		&mut self, duration_since_epoch: Duration,
+	) -> Result<(Offer, bool), ()> {
+		self.prune_expired_offers(duration_since_epoch, false);
+
+		// Find the freshest unused offer, where "freshness" is based on when the invoice was confirmed
+		// persisted by the server
+		let newest_ready_offer_opt = self
+			.offers_with_idx()
+			.filter_map(|(idx, offer)| match offer.status {
+				OfferStatus::Ready { invoice_confirmed_persisted_at } => {
+					Some((idx, offer, invoice_confirmed_persisted_at))
+				},
+				_ => None,
+			})
+			.max_by(|a, b| a.2.cmp(&b.2))
+			.map(|(idx, offer, _)| (idx, offer.offer.clone()));
+		if let Some((idx, newest_ready_offer)) = newest_ready_offer_opt {
+			self.offers[idx].as_mut().map(|offer| offer.status = OfferStatus::Used);
+			return Ok((newest_ready_offer, true));
+		}
+
+		// If no unused offers are available, return the used offer with the latest absolute expiry
+		self.offers_with_idx()
+			.filter(|(_, offer)| matches!(offer.status, OfferStatus::Used))
+			.max_by(|a, b| {
+				let abs_expiry_a = a.1.offer.absolute_expiry().unwrap_or(Duration::MAX);
+				let abs_expiry_b = b.1.offer.absolute_expiry().unwrap_or(Duration::MAX);
+				abs_expiry_a.cmp(&abs_expiry_b)
+			})
+			.map(|(_, cache_offer)| (cache_offer.offer.clone(), false))
+			.ok_or(())
+	}
+
+	/// Remove expired offers from the cache, returning whether new offers are needed.
+	pub(super) fn prune_expired_offers(
+		&mut self, duration_since_epoch: Duration, timer_tick_occurred: bool,
+	) -> bool {
+		// Remove expired offers from the cache.
+		let mut offer_was_removed = false;
+		for offer_opt in self.offers.iter_mut() {
+			let offer_is_expired = offer_opt
+				.as_ref()
+				.map_or(false, |offer| offer.offer.is_expired_no_std(duration_since_epoch));
+			if offer_is_expired {
+				offer_opt.take();
+				offer_was_removed = true;
+			}
+		}
+
+		// Allow more offer paths requests to be sent out in a burst roughly once per minute, or if an
+		// offer was removed.
+		if timer_tick_occurred || offer_was_removed {
+			self.reset_offer_paths_request_attempts()
+		}
+
+		self.needs_new_offer_idx(duration_since_epoch).is_some()
+			&& self.offer_paths_request_attempts < MAX_UPDATE_ATTEMPTS
+	}
+
+	/// Returns whether the new paths we've just received from the static invoice server should be used
+	/// to build a new offer.
+	pub(super) fn should_build_offer_with_paths(
+		&self, offer_paths: &[BlindedMessagePath], offer_paths_absolute_expiry_secs: Option<u64>,
+		duration_since_epoch: Duration,
+	) -> bool {
+		if self.needs_new_offer_idx(duration_since_epoch).is_none() {
+			return false;
+		}
+
+		// Require the offer that would be built using these paths to last at least
+		// MIN_OFFER_PATHS_RELATIVE_EXPIRY_SECS.
```

Comment:
I think constants in comments can be made into links. There may be other instances.